#VRML_SIM R2025a utf8

WorldInfo {
  info [
    "A model of the NASA Sojourner Rover on Mars."
  ]
  title "Sojourner Rover"
  gravity 3.73
  basicTimeStep 16
  lineScale 0.2
  contactProperties [
    ContactProperties {
      maxContactJoints 21
    }
  ]
}
Viewpoint {
  fieldOfView 0.895398
  orientation -0.1656851679088041 -0.026591406942797106 0.985820126702551 3.701317370049612
  position -295866.82293736073 -224740.64474379364 -198753.65014182168
  near 0.1
}
Background {
  skyColor [
    0.5 0.2 0
  ]
}
DirectionalLight {
  direction 0.1 -0.5 0.3
}
Fog {
  color 0.5 0.2 0
  fogType "EXPONENTIAL"
  visibilityRange 800
}
DEF GROUND Solid {
  children [
    DEF GROUND_SHAPE Shape {
      appearance PBRAppearance {
        baseColor 0.8 0.2 0
        roughness 1
        metalness 0
      }
      geometry Plane {
        size 100 100
      }
    }
  ]
  name "ground"
  boundingObject USE GROUND_SHAPE
  locked TRUE
}
Robot {
  translation 1.5147896239640941 5.412476011017817e-07 0.20161657467440627
  rotation 0.003032325895320208 -0.9999454883119414 0.009991266399085505 0.008244159921923686
  children [
    Camera {
      translation 0.2 0 0.1
      rotation 0 1 0 0
      width 160
      height 120
    }
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0.8 0.6 0.2
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.4 0.3 0.2
      }
    }
    DEF WHEEL_FL HingeJoint {
      jointParameters HingeJointParameters {
        position 55.088264944421105
        axis 0 1 0
        anchor 0.2 0.2 -0.1
      }
      device [
        RotationalMotor {
          name "FrontLeftWheel"
        }
      ]
      endPoint Solid {
        translation 0.2 0.2 -0.1
        rotation 0 1 0 4.822782486984414
        children [
          DEF WHEEL_SHAPE Shape {
            appearance PBRAppearance {
              baseColor 0.3 0.3 0.3
            }
            geometry Cylinder {
              height 0.05
              radius 0.1
            }
          }
        ]
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density 100
        }
        linearVelocity -0.03954812231107108 -1.9553161475999598e-06 0.013480329928281558
        angularVelocity -1.4265662949052983e-05 0.3988160732557163 -1.1157040834555708e-05
      }
    }
    DEF WHEEL_FR HingeJoint {
      jointParameters HingeJointParameters {
        position 55.08826462133717
        axis 0 1 0
        anchor 0.2 -0.2 -0.1
      }
      device [
        RotationalMotor {
          name "FrontRightWheel"
        }
      ]
      endPoint Solid {
        translation 0.2 -0.2 -0.1
        rotation 0 1 0 4.8227821639004755
        children [
          USE WHEEL_SHAPE
        ]
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density 100
        }
        linearVelocity -0.03955864670221757 -1.7589934477890993e-06 0.013471468010093513
        angularVelocity -1.4132641262137218e-05 0.3988153554306823 -1.0545475838418782e-05
      }
    }
    DEF WHEEL_ML HingeJoint {
      jointParameters HingeJointParameters {
        position 55.086405549707635
        axis 0 1 0
        anchor 0 0.2 -0.1
      }
      device [
        RotationalMotor {
          name "MiddleLeftWheel"
        }
      ]
      endPoint Solid {
        translation 0 0.2 -0.1
        rotation 0 1 0 4.820923092270943
        children [
          USE WHEEL_SHAPE
        ]
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density 100
        }
        linearVelocity -0.039530119391328125 -1.3642424755889156e-06 0.009194967480185189
        angularVelocity -1.4425614667446319e-05 0.3988215821587541 -1.0844114511107702e-05
      }
    }
    DEF WHEEL_MR HingeJoint {
      jointParameters HingeJointParameters {
        position 55.08640621192442
        axis 0 1 0
        anchor 0 -0.2 -0.1
      }
      device [
        RotationalMotor {
          name "MiddleRightWheel"
        }
      ]
      endPoint Solid {
        translation 0 -0.2 -0.1
        rotation 0 1 0 4.820923754487728
        children [
          USE WHEEL_SHAPE
        ]
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density 100
        }
        linearVelocity -0.0395383841537632 -1.3648076316012262e-06 0.009187474126924371
        angularVelocity -1.4425635115778611e-05 0.3988215535692222 -1.084412323797918e-05
      }
    }
    DEF WHEEL_BL HingeJoint {
      jointParameters HingeJointParameters {
        position 55.063740607331674
        axis 0 1 0
        anchor -0.2 0.2 -0.1
      }
      device [
        RotationalMotor {
          name "BackLeftWheel"
        }
      ]
      endPoint Solid {
        translation -0.2 0.2 -0.1
        rotation 0 1 0 4.798258149894982
        children [
          USE WHEEL_SHAPE
        ]
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density 100
        }
        linearVelocity -0.03948956975142645 5.347834937601608e-06 0.0050080555126218946
        angularVelocity -1.4170437611629574e-05 0.39882463342012386 -1.0761500312594001e-05
      }
    }
    DEF WHEEL_BR HingeJoint {
      jointParameters HingeJointParameters {
        position 55.06374207488222
        axis 0 1 0
        anchor -0.2 -0.2 -0.1
      }
      device [
        RotationalMotor {
          name "BackRightWheel"
        }
      ]
      endPoint Solid {
        translation -0.2 -0.2 -0.1
        rotation 0 1 0 4.798259617445529
        children [
          USE WHEEL_SHAPE
        ]
        boundingObject USE WHEEL_SHAPE
        physics Physics {
          density 100
        }
        linearVelocity -0.03949771171740183 5.368873975176961e-06 0.005000589157754339
        angularVelocity -1.4168297478704957e-05 0.39882460175671375 -1.0761554663472352e-05
      }
    }
  ]
  name "Sojourner"
  boundingObject USE BODY
  physics Physics {
    density 500
  }
  controller "rover_controller"
  linearVelocity -0.04164419828129399 -1.3111551982027216e-06 0.009206981900338926
  angularVelocity 2.0397735434793517e-05 -0.021164313648672068 -2.101893680862532e-05
}
Solid {
  translation 2 0 0.3
  children [
    DEF ROCK_SHAPE Shape {
      appearance PBRAppearance {
        baseColor 0.8 0.35 0.3
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.5 0.5 0.5
      }
    }
  ]
  name "rock1"
  boundingObject USE ROCK_SHAPE
}
Solid {
  translation -2 2 0.3
  children [
    USE ROCK_SHAPE
  ]
  name "rock2"
  boundingObject USE ROCK_SHAPE
}
Solid {
  translation -2 -2 0.3
  children [
    USE ROCK_SHAPE
  ]
  name "rock3"
  boundingObject USE ROCK_SHAPE
}
Robot {
  controller "world_supervisor"
  supervisor TRUE
}
